sessiontxn/txn.go

```diff
@@ -67,3 +70,56 @@ func CanReuseTxnWhenExplicitBegin(sctx sessionctx.Context) bool {
 	// initialized with snapshot ts.
 	return txnCtx.History == nil && !txnCtx.IsStaleness && sessVars.SnapshotTS == 0
 }
+
+// IsTxnRetryable (if returns true) means the transaction could retry.
+// If the transaction is in pessimistic mode, do not retry.
+// If the session is already in transaction, enable retry or internal SQL could retry.
+// If not, the transaction could always retry, because it should be auto committed transaction.
+// Anyway the retry limit is 0, the transaction could not retry.
+func IsTxnRetryable(sessVars *variable.SessionVars) bool {
```

Comment:
It will only be used by optimistic txn, so it can be moved to isolation package and can be private